File: extensions/pi-looper/automations/pr-branch-update-decision.py
#!/usr/bin/env python3
"""Decide whether a PR branch can be updated mechanically or needs one worker."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Any


class CommandError(RuntimeError):
    pass


def run_git(repo: str, args: list[str], check: bool = True) -> subprocess.CompletedProcess[str]:
    result = subprocess.run(
        ["git", "-C", repo, *args],
        text=True,
        capture_output=True,
        check=False,
    )
    if check and result.returncode != 0:
        raise CommandError((result.stderr or result.stdout or f"git failed: {args!r}").strip())
    return result


def count_commits(repo: str, rev_range: str) -> int:
    output = run_git(repo, ["rev-list", "--count", rev_range]).stdout.strip()
    return int(output or "0")


def merge_tree_is_clean(repo: str, head_ref: str, base_ref: str) -> bool:
    result = run_git(repo, ["merge-tree", "--write-tree", head_ref, base_ref], check=False)
    return result.returncode == 0


def decide(ahead: int, behind: int, conflict_free: bool) -> dict[str, Any]:
    diverged = ahead > 0 and behind > 0
    if behind <= 0:
        action = "no_update"
        reason = "head_contains_base"
    elif conflict_free:
        action = "mechanical_update"
        reason = "fast_forward" if ahead == 0 else "clean_merge"
    else:
        action = "delegate_worker"
        reason = "merge_conflict"

    return {
        "action": action,
        "reason": reason,
        "ahead": ahead,
        "behind": behind,
        "diverged": diverged,
        "conflictFree": conflict_free,
    }


def decide_live(repo: str, head_ref: str, base_ref: str) -> dict[str, Any]:
    run_git(repo, ["rev-parse", "--verify", head_ref])
    run_git(repo, ["rev-parse", "--verify", base_ref])
    ahead = count_commits(repo, f"{base_ref}..{head_ref}")
    behind = count_commits(repo, f"{head_ref}..{base_ref}")
    conflict_free = True if behind <= 0 else merge_tree_is_clean(repo, head_ref, base_ref)
    decision = decide(ahead=ahead, behind=behind, conflict_free=conflict_free)
    decision.update({"headRef": head_ref, "baseRef": base_ref})
    return decision


def decide_fixture(path: str) -> dict[str, Any]:
    data = json.loads(Path(path).read_text())
    return decide(
        ahead=int(data.get("ahead", 0)),
        behind=int(data.get("behind", 0)),
        conflict_free=bool(data.get("conflictFree", True)),
    )


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", help="Repository path for live git checks.")
    parser.add_argument("--head", help="PR head ref, for example origin/my-branch.")
    parser.add_argument("--base", help="Base ref, for example origin/main.")
    parser.add_argument("--fixture", help="Fixture JSON with ahead, behind, and conflictFree.")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if args.fixture:
        decision = decide_fixture(args.fixture)
    else:
        if not args.repo or not args.head or not args.base:
            raise ValueError("--repo, --head, and --base are required unless --fixture is used")
        decision = decide_live(args.repo, args.head, args.base)
    print(json.dumps(decision, ensure_ascii=False, sort_keys=True))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"pr-branch-update-decision.py: {error}", file=sys.stderr)
        raise SystemExit(2)

File: extensions/deadloop/automations/pr-branch-update-decision.ts
#!/usr/bin/env node
// Decide whether a PR branch can be updated mechanically or needs one worker.
// CommonJS-shaped so it can run directly with `node pr-branch-update-decision.ts`.

const fs = require("node:fs") as typeof import("node:fs");
const { spawnSync } = require("node:child_process") as typeof import("node:child_process");

type BranchUpdateDecision = Record<string, any>;

function runGitForBranchUpdate(repo: string, args: string[], options: { check?: boolean } = {}) {
  const result = spawnSync("git", ["-C", repo, ...args], { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] });
  if (options.check !== false && result.status !== 0) {
    throw new Error((result.stderr || result.stdout || `git failed: ${args.join(" ")}`).trim());
  }
  return result;
}

function countCommitsForBranchUpdate(repo: string, revRange: string): number {
  const output = runGitForBranchUpdate(repo, ["rev-list", "--count", revRange]).stdout.trim();
  return Number(output || 0);
}

function mergeTreeIsClean(repo: string, headRef: string, baseRef: string): boolean {
  return runGitForBranchUpdate(repo, ["merge-tree", "--write-tree", headRef, baseRef], { check: false }).status === 0;
}

function worktreeIsClean(repo: string): boolean {
  return runGitForBranchUpdate(repo, ["status", "--short"]).stdout.trim() === "";
}

function revParseForBranchUpdate(repo: string, ref: string): string {
  return runGitForBranchUpdate(repo, ["rev-parse", "--verify", ref]).stdout.trim();
}

function decideBranchUpdate(
  ahead: number,
  behind: number,
  conflictFree: boolean,
  cleanWorktree = true,
  headMatchesExpected = true,
): BranchUpdateDecision {
  const diverged = ahead > 0 && behind > 0;
  let action: string;
  let reason: string;
  if (!cleanWorktree) {
    action = "blocked";
    reason = "dirty_worktree";
  } else if (!headMatchesExpected) {
    action = "blocked";
    reason = "stale_head";
  } else if (behind <= 0) {
    action = "no_update";
    reason = "head_contains_base";
  } else if (conflictFree) {
    action = "mechanical_update";
    reason = ahead === 0 ? "fast_forward" : "clean_merge";
  } else {
    action = "delegate_worker";
    reason = "merge_conflict";
  }

  return {
    action,
    reason,
    ahead,
    behind,
    diverged,
    conflictFree,
    cleanWorktree,
    headMatchesExpected,
  };
}

function decideBranchUpdateLive(repo: string, headRef: string, baseRef: string, expectedHeadRef: string | undefined): BranchUpdateDecision {
  const headOid = revParseForBranchUpdate(repo, headRef);
  runGitForBranchUpdate(repo, ["rev-parse", "--verify", baseRef]);
  const expectedHeadOid = expectedHeadRef ? revParseForBranchUpdate(repo, expectedHeadRef) : headOid;
  const ahead = countCommitsForBranchUpdate(repo, `${baseRef}..${headRef}`);
  const behind = countCommitsForBranchUpdate(repo, `${headRef}..${baseRef}`);
  const conflictFree = behind <= 0 || mergeTreeIsClean(repo, headRef, baseRef);
  return {
    ...decideBranchUpdate(ahead, behind, conflictFree, worktreeIsClean(repo), headOid === expectedHeadOid),
    headRef,
    baseRef,
    headOid,
    expectedHeadRef: expectedHeadRef ?? null,
    expectedHeadOid,
  };
}

function boolFromFixtureForBranchUpdate(value: unknown, defaultValue: boolean): boolean {
  return value === undefined ? defaultValue : Boolean(value);
}

function decideBranchUpdateFixture(file: string): BranchUpdateDecision {
  const data = JSON.parse(fs.readFileSync(file, "utf8"));
  return decideBranchUpdate(
    Number(data.ahead || 0),
    Number(data.behind || 0),
    boolFromFixtureForBranchUpdate(data.conflictFree, true),
    boolFromFixtureForBranchUpdate(data.cleanWorktree, true),
    boolFromFixtureForBranchUpdate(data.headMatchesExpected, true),
  );
}

function requiredBranchUpdateValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1];
  if (!value || value.startsWith("--")) throw new Error(`${flag} requires a value`);
  return value;
}

function parseBranchUpdateArgs(argv: string[]): BranchUpdateDecision {
  const parsed: BranchUpdateDecision = {};
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--help" || token === "-h") {
      parsed.help = true;
      continue;
    }
    if (["--repo", "--head", "--base", "--fixture", "--expected-head-ref"].includes(token)) {
      const key = token.slice(2).replace(/-([a-z])/g, (_match, char) => char.toUpperCase());
      parsed[key] = requiredBranchUpdateValue(argv, index, token);
      index += 1;
      continue;
    }
    throw new Error(`unknown flag: ${token}`);
  }
  return parsed;
}

function branchUpdateHelp(): string {
  return [
    "Usage: pr-branch-update-decision.ts --fixture FILE",
    "   or: pr-branch-update-decision.ts --repo PATH --head REF --base REF [--expected-head-ref REF]",
  ].join("\n");
}

function main(argv: string[] = process.argv.slice(2)): number {
  const args = parseBranchUpdateArgs(argv);
  if (args.help) {
    process.stdout.write(`${branchUpdateHelp()}\n`);
    return 0;
  }
  const decision = args.fixture
    ? decideBranchUpdateFixture(args.fixture)
    : (() => {
        if (!args.repo || !args.head || !args.base) throw new Error("--repo, --head, and --base are required unless --fixture is used");
        return decideBranchUpdateLive(args.repo, args.head, args.base, args.expectedHeadRef);
      })();
  process.stdout.write(`${JSON.stringify(decision)}\n`);
  return 0;
}

if (require.main === module) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`pr-branch-update-decision.ts: ${error instanceof Error ? error.message : String(error)}`);
    process.exitCode = 2;
  }
}

module.exports = { decideBranchUpdate, decideBranchUpdateFixture, decideBranchUpdateLive };
